packages/squiggle-lang/src/fr/common.ts

```diff
@@ -107,13 +107,21 @@ myFn = typeOf({|e| e})`,
         ],
         frOr(frAny({ genericName: "A" }), frAny({ genericName: "B" })),
         ([fn, fallbackFn], reducer) => {
+          const stackSize = reducer.stack.size();
+          const frameStackSize = reducer.frameStack.frames.length;
+
           try {
             return { tag: "1", value: reducer.call(fn, []) };
           } catch (e) {
             if (!(e instanceof ErrorMessage)) {
               // This doesn't looks like an error in user code, treat it as fatal
               throw e;
             }
+
+            reducer.stack.shrink(stackSize);
+            while (reducer.frameStack.frames.length > frameStackSize) {
+              reducer.frameStack.pop();
+            }
```

Comment:
This is not the clearest possible solution, slightly fragile, but good enough for now